runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -700,25 +718,35 @@ Java_java_lang_invoke_MethodHandleNatives_resolve(JNIEnv *env, jclass clazz, job
 			/* Refetch reference after GC point */
 			membernameObject = J9_JNI_UNWRAP_REFERENCE(self);
 			nameObject = J9VMJAVALANGINVOKEMEMBERNAME_NAME(currentThread, membernameObject);
-			name = vmFuncs->copyStringToUTF8WithMemAlloc(currentThread, nameObject, J9_STR_NULL_TERMINATE_RESULT, "", 0, NULL, 0, &nameLength);
+			name = vmFuncs->copyStringToJ9UTF8WithMemAlloc(currentThread, nameObject, 0, "", 0, NULL, 0);
 			if (NULL == name) {
 				vmFuncs->setNativeOutOfMemoryError(currentThread, 0, 0);
 				goto done;
 			}
 
-			Trc_JCL_java_lang_invoke_MethodHandleNatives_resolve_NAS(env, name, signature);
+			j9object_t callerObject = (NULL == caller) ? NULL : J9_JNI_UNWRAP_REFERENCE(caller);
+			J9Class *callerClass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, callerObject);
 
-			if (J9_ARE_ANY_BITS_SET(flags, MN_IS_METHOD | MN_IS_CONSTRUCTOR)) {
-				j9object_t callerObject = (NULL == caller) ? NULL : J9_JNI_UNWRAP_REFERENCE(caller);
-				J9Class *callerClass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, callerObject);
+			Trc_JCL_java_lang_invoke_MethodHandleNatives_resolve_NAS(env, J9UTF8_LENGTH(name), J9UTF8_DATA(name), J9UTF8_LENGTH(signature), J9UTF8_DATA(signature));
 
-				J9JNINameAndSignature nas;
-				UDATA lookupOptions = J9_LOOK_JNI;
+			if (J9_ARE_ANY_BITS_SET(flags, MN_IS_METHOD | MN_IS_CONSTRUCTOR)) {
+				J9NameAndSignature nas;
+				UDATA lookupOptions = J9_LOOK_DIRECT_NAS;
 
 				if (JNI_TRUE == speculativeResolve) {
 					lookupOptions |= J9_LOOK_NO_THROW;
 				}
 
+#if JAVA_SPEC_VERSION >= 16
+				if ((NULL != callerClass) && J9_ARE_NO_BITS_SET(lookupMode, MN_UNCONDITIONAL_MODE)) {
+					lookupOptions |= J9_LOOK_CLCONSTRAINTS;
```

Comment:
It's surprising to skip class loader constraints.  Can you add a comment here to explain why UNCONDITIONAL means clconstraints don't need to be checked?

Reply:
The comment that was added describes in english the `if` statement, but doesn't expand on the **why**.  The **why** is critical here and will help to ensure the behaviour is correct and consistent

Reply:
updated comment